sdk/identity/Azure.Identity/src/ClientCertificateCredential.cs

```diff
@@ -135,5 +183,215 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
+                    default:
+                        throw new CredentialUnavailableException("Only .pfx and .pem files are supported.");
+                }
+            }
+
+            private const int BufferSize = 4 * 1024;
+
+            private async Task<X509Certificate2> LoadCertificateFromPfxFileAsync(bool async, string clientCertificatePath, CancellationToken cancellationToken)
+            {
+                if (!(Certificate is null))
+                {
+                    return Certificate;
+                }
+
+                try
+                {
+                    if (!async)
+                    {
+                        Certificate = new X509Certificate2(clientCertificatePath);
+                    }
+                    else
+                    {
+                        List<byte> certContents = new List<byte>();
+                        byte[] buf = new byte[BufferSize];
+                        int offset = 0;
+                        using (Stream s = File.OpenRead(clientCertificatePath))
+                        {
+                            while (true)
+                            {
+                                int read = await s.ReadAsync(buf, offset, buf.Length, cancellationToken).ConfigureAwait(false);
+                                for (int i = 0; i < read; i++)
+                                {
+                                    certContents.Add(buf[i]);
+                                }
+
+                                if (read == 0)
+                                {
+                                    break;
+                                }
+                            }
+                        }
+
+                        Certificate = new X509Certificate2(certContents.ToArray());
+                    }
+
+                    return Certificate;
+                }
+                catch (Exception e) when (!(e is OperationCanceledException))
+                {
+                    throw new CredentialUnavailableException("Could not load certificate file", e);
+                }
+            }
+
+            private async Task<X509Certificate2> LoadCertificateFromPemFileAsync(bool async, string clientCertificatePath, CancellationToken cancellationToken)
+            {
+                if (!(Certificate is null))
+                {
+                    return Certificate;
+                }
+
+                string certficateText;
+
+                try
+                {
+                    if (!async)
+                    {
+                        certficateText = File.ReadAllText(clientCertificatePath);
+                    }
+                    else
+                    {
+                        StringBuilder sb = new StringBuilder();
+                        char[] buf = new char[BufferSize];
+                        int offset = 0;
+
+                        using (StreamReader sr = new StreamReader(clientCertificatePath))
```

Comment:
https://docs.microsoft.com/en-us/dotnet/api/system.io.streamreader.readtoendasync?view=netframework-4.8 ?

Reply:
There's no overload that takes a cancellation token, so I open coded this so I could check for cancellation.  Do you think that's overkill?

Reply:
Might be a bit of overkill considering it's a small local file

Reply:
SGTM - Will move then.